DRA API: graduation to GA

pohly opened the pull request:
#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:

Graduation of DRA Structured Parameters (= core DRA) to GA.

#### Which issue(s) this PR is related to:

Fixes: https://github.com/kubernetes/kubernetes/issues/129891
https://github.com/kubernetes/kubernetes/issues/131903
KEP: https://github.com/kubernetes/enhancements/issues/4381

#### Special notes for your reviewer:

TODO:
- [x] feature gate
- [x] use v1 API everywhere
- [x] E2E and integration test for v1beta2

[GA criteria](https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/4381-dra-structured-parameters#ga):
- [x] At least 3 real-world driver implementations from different organizations: Intel, Google, Nvidia, ...
- [x] Standardization of at least one attribute suitable for intra-node topology alignment across drivers (e.g., NUMA or PCIe root, as determined to be appropriate and useful): [PCI Root](https://github.com/kubernetes/enhancements/pull/5316)
- [x] Allowing time for feedback: in beta since 1.32, plenty of discussions in WG Device Management
- [x] Scalability and performance analysis done and published: [WIP Google Doc](https://docs.google.com/document/d/1YzPR3j0gxDNwL6iaY28trUsKVb2xyhrwHY-DqVcVaUE/edit?tab=t.0)
- [x] Define kubelet restart, node bootstrap/restart, driver runtime, and upgrade/downgrade failure modes: https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/pluginmanager/pluginwatcher/README.md
- [x] Implement kubelet, [DRA kubeletplugin framework](https://github.com/kubernetes/kubernetes/tree/master/staging/src/k8s.io/dynamic-resource-allocation/kubeletplugin), or [example driver](https://github.com/kubernetes-sigs/dra-example-driver) changes based on above analysis:
  - [x] [seamless upgrades](https://github.com/kubernetes/kubernetes/pull/129832)
  - [x] [kubelet connection monitoring](https://github.com/kubernetes/kubernetes/pull/132058)
  - [x] [liveness probe in example driver](https://github.com/kubernetes-sigs/dra-example-driver/pull/104)
  - [x] automated upgrade/downgrade testing: https://testgrid.k8s.io/sig-node-dynamic-resource-allocation#ci-dra-integration (= https://github.com/kubernetes/kubernetes/tree/master/test/e2e_dra)
- [ ]    Address existing in-flight work and bugs ex.
        ~~https://github.com/kubernetes/kubernetes/pull/129799~~: solved via https://github.com/kubernetes/kubernetes/pull/132058,
        https://github.com/kubernetes/kubernetes/issues/128964,
        https://github.com/kubernetes/kubernetes/issues/129402 (replaced by https://github.com/kubernetes/website/issues/51012).:
        https://github.com/kubernetes/kubernetes/pull/133113
        https://github.com/orgs/kubernetes/projects/95/views/10


#### Does this PR introduce a user-facing change?
```release-note
Dynamic Resource Allocation: graduated core functionality to general availability (GA). This newly stable feature uses the _structured parameters_ flavor of DRA.
```

pohly committed:
DRA API: verbatim copy v1beta2 -> v1

pohly committed:
DRA API: s/v1beta2/v1/ and generated files

pohly committed:
DRA API: v1 registration + tests

pohly committed:
DRA: graduate DynamicResourceAllocation feature to GA

It hasn't been on-by-default before, therefore it does not get locked to the
new default on yet. This has some impact on the scheduler configuration
because the plugin is now enabled by default.

Because the feature is now GA, it doesn't need to be a label on E2E tests,
which wouldn't be possible anyway once it gets removed entirely.

pohly committed:
DRA: use v1 API

As before when adding v1beta2, DRA drivers built using the
k8s.io/dynamic-resource-allocation helper packages remain compatible with all
Kubernetes release >= 1.32. The helper code picks whatever API version is
enabled from v1beta1/v1beta2/v1.

However, the control plane now depends on v1, so a cluster configuration where
only v1beta1 or v1beta2 are enabled without the v1 won't work.

lmktfy commented:
Change llog suggestion
```diff
-Dynamic Resource Allocation (DRA): the core functionality (also known as "structured parameters") is GA.
+Dynamic Resource Allocation: graduated core functionality to general availability (GA). This newly stable feature uses the _structured parameters_ flavor of DRA.
```
?

pohly commented:
Ack, that sounds better:

/release-note-edit
```release-note
+Dynamic Resource Allocation: graduated core functionality to general availability (GA). This newly stable feature uses the _structured parameters_ flavor of DRA.
```

pohly commented:
/test pull-kubernetes-unit

"TestPatchResourceTimeout"

pohly commented:
/skip

apidiff: client-go changes because of the v1
windows-master: unrelated flakes

pohly commented:
/test pull-kubernetes-integration

Job stuck, OOM?

pohly commented:
/test pull-kubernetes-e2e-gce

pohly commented:
/test pull-kubernetes-unit

Flake.

macsko commented on cmd/kube-scheduler/app/server_test.go:
```diff
@@ -249,13 +250,18 @@ leaderElection:
 			wantPlugins: map[string]*config.Plugins{
 				"default-scheduler": func() *config.Plugins {
 					plugins := defaults.ExpandedPluginsV1.DeepCopy()
+					// With this (and only this?!) config comes DynamicResources after DefaultPreemption.
+					plugins.PreEnqueue.Enabled[1], plugins.PreEnqueue.Enabled[2] = plugins.PreEnqueue.Enabled[2], plugins.PreEnqueue.Enabled[1]
+					plugins.PostFilter.Enabled[0], plugins.PostFilter.Enabled[1] = plugins.PostFilter.Enabled[1], plugins.PostFilter.Enabled[0]
```
Why are you modifying the plugins' order this way, but not explicitly as below?
```suggestion
					plugins.PreEnqueue.Enabled = []config.Plugin{
						...
					}
					plugins.PostFilter.Enabled = []config.Plugin{
						...
					}
```

pohly replied:
If you prefer to avoid swapping, then I can also do that, for one or the other config variant.

I chose swapping to highlight that only the order of the entries from `defaults.ExpandedPluginsV1` is different, but not the entries.

macsko reviewed (commented):
Looks good from the scheduler's perspective, but deferring to @sanposhiho for additional review.

micahhausler commented:
/assign @ritazh 
To review RBAC rules

pohly commented on pkg/scheduler/eventhandlers_test.go:
```diff
@@ -529,8 +529,8 @@ func TestAddAllEventHandlers(t *testing.T) {
-				"daemonsets.v1.apps":    fwk.Add | fwk.Delete,
-				"custommetrics.v1beta1": fwk.Update,
+				"daemonsets.v1.apps": fwk.Add | fwk.Delete,
+				"custommetrics.v1":   fwk.Update,
```
Unintended! Thanks.

I wonder why this doesn't fail the tests. Ah, perhaps because it is the "one illegal GVK form" and ignored either way. Still, I'll revert this.

pohly commented on staging/src/k8s.io/dynamic-resource-allocation/api/v1beta2/doc.go:
```diff
@@ -0,0 +1,24 @@
+limitations under the License.
+*/
+
+// k8s:conversion-gen specifies the latest API version in k8s.io/api/resource.
+//
+// +k8s:conversion-gen=k8s.io/api/resource/v1beta2
+// +k8s:conversion-gen-external-types=k8s.io/api/resource/v1
```
Also in the v1beta1 conversion.

pohly commented on staging/src/k8s.io/dynamic-resource-allocation/api/v1beta2/conversion_test.go:
```diff
@@ -0,0 +1,337 @@
+			},
+		},
+		{
+			name: "v1beta2 to latest with alternatives",
```
Also fixed in v1beta1 conversion.

pohly commented on test/integration/dra/dra_test.go:
```diff
@@ -219,25 +226,10 @@ func TestDRA(t *testing.T) {
 				tCtx.Run("ResourceClaimDeviceStatus", func(tCtx ktesting.TContext) { testResourceClaimDeviceStatus(tCtx, false) })
 			},
 		},
-		"core": {
```
Yes.

pohly commented:
> Will this PR close https://github.com/kubernetes/kubernetes/issues/129891?

Yes. Added to PR description.

pohly commented on staging/src/k8s.io/dynamic-resource-allocation/structured/allocator.go:
```diff
@@ -1272,26 +1258,24 @@ func (alloc *allocator) createNodeSelector(result []internalDeviceResult) (*v1.N
 		var nodeSelector *v1.NodeSelector
 		if ptr.Deref(slice.Spec.PerDeviceNodeSelection, false) {
-			if result[i].basic.NodeName != nil {
-				nodeName = draapi.MakeUniqueString(*result[i].basic.NodeName)
```
I hadn't, but that's a good point. I found that I had to update the scheduler_perf configuration because disabling queueing hints is not possible in combination with the 1.34 v1 API.

When comparing master+that change against the full branch I don't see significant differences in the results.

pohly commented on api/discovery/apis__admissionregistration.k8s.io__v1beta1.json:
ValidatingAdmissionPolicyBinding in v1beta2 is supposed to be removed in 1.34, same for the other types. "make verify" must be using some kind of hack to prevent verification failures. I see the same removal in other PRs which re-generate API files.

It probably should get removed in a separate PR.

pohly committed:
DRA API: bump storage version to v1beta2

This avoids the overhead for the more complex conversion to v1beta1 and might
make it a bit more realistic to get rid of the v1beta1 eventually.

The expected GVK must be set explicitly because when emulating 1.33,
v1beta1 is the default although the fixed storage version is v1beta2.

pohly commented on pkg/apis/resource/install/install.go:
```diff
@@ -38,7 +39,8 @@ func Install(scheme *runtime.Scheme) {
 	// TODO(https://github.com/kubernetes/kubernetes/issues/129889) We should
 	// change the serialization version to v1beta2 for 1.34.
-	utilruntime.Must(scheme.SetVersionPriority(v1beta1.SchemeGroupVersion, v1beta2.SchemeGroupVersion, v1alpha3.SchemeGroupVersion))
+	utilruntime.Must(scheme.SetVersionPriority(v1beta1.SchemeGroupVersion, v1beta2.SchemeGroupVersion, v1.SchemeGroupVersion, v1alpha3.SchemeGroupVersion))
```
Included. I asked on Slack about the right way to do this. Version emulation makes it a bit more complex.

sanposhiho reviewed (approved):
/approve

sig-scheduling

seans3 commented:
/remove-sig api-machinery

ritazh commented:
/lgtm

sig-auth

pohly commented:
/skip

[pull-kubernetes-apidiff-client-go](https://prow.k8s.io/view/gs/kubernetes-ci-logs/pr-logs/pull/132706/pull-kubernetes-apidiff-client-go/1945748211817779200) is complaining about the inevitable client-go API changes (new method in interface).

pohly committed:
integration: use --runtime-config-emulation-forward-compatible

Some tests do version emulation and need the DRA feature. In that combination
the --runtime-config-emulation-forward-compatible option is needed to allow
enabling the V1 API although it's only available in 1.34.

pohly commented:
I was watching pull-kubernetes-integration while it was running and it seemed to be fine, but now it has died again without reporting any status. Will investigate more tomorrow...

pohly commented:
pull-kubernetes-integration is giving me trouble, it just dies after a time out without uploading logs or any results. All of the tests pass for me locally so far.

I'm putting this back into draft mode while I debug this. The PR itself can still be reviewed.

pohly commented:
/test pull-kubernetes-integration pull-kubernetes-verify

pohly commented:
/test pull-kubernetes-integration pull-kubernetes-verify

pohly commented:
> Upgrade/downgrade testing breaks in the "DRA: use v1 API". Before, when the apiserver gets restarted, clients detect that the watch has died and retry with List+Watch. 

It's [working in the CI](https://prow.k8s.io/view/gs/kubernetes-ci-logs/pr-logs/pull/132706/pull-kubernetes-dra-integration/1946319331021295616). Still investigating...

pohly commented:
/test pull-kubernetes-dra-integration

bart0sh commented:
/test pull-kubernetes-unit-windows-master

pohly commented:
/skip

pohly commented on pkg/controller/devicetainteviction/device_taint_eviction.go:
```diff
@@ -165,11 +165,7 @@ func (p pool) getTaintedDevices() []taintedDevice {
-				continue
-			}
-			for _, taint := range device.Basic.Taints {
+			for _, taint := range device.Taints {
```
The manual changes required for v1beta1 to v1beta2/v1 are:
- removal of `Basic` indirection in ResourceSlice
- adding `Exactly` indirection in requests

With hindsight, this is something that could have been done in 1.33 - but it's still fairly mechanical and failed in obvious ways if not done (compile error).

liggitt commented on pkg/apis/resource/install/install.go:
```diff
@@ -38,7 +39,6 @@ func Install(scheme *runtime.Scheme) {
-	// change the serialization version to v1beta2 for 1.34.
-	utilruntime.Must(scheme.SetVersionPriority(v1beta1.SchemeGroupVersion, v1beta2.SchemeGroupVersion, v1alpha3.SchemeGroupVersion))
+	utilruntime.Must(v1.AddToScheme(scheme))
+	utilruntime.Must(scheme.SetVersionPriority(v1beta2.SchemeGroupVersion, v1.SchemeGroupVersion, v1beta1.SchemeGroupVersion, v1alpha3.SchemeGroupVersion))
```
@Jefftree can you give guidance here, now that emulated version / compatibility version storage wiring auto-lowers the storage version for a particular resource to whatever the compatibility version (n-1) server can handle.

Ideally, we could just put the most stable to least stable order in static install.go files, but that may introduce new instant-break situations when we create new release branches and master gets a next-minor development tag compared to the release branch, but must still pass tests for the previous minor

liggitt replied:
I *think* just the first version in this list really matters.

Putting v1beta2 first makes sense to me for 1.34. Updating expectedGVK to "v1beta2" in etcd/data integration test as a result is expected.

Add a TODO comment here to prefer v1 in 1.35.

Jefftree replied:
> @Jefftree can you give guidance here, now that emulated version / compatibility version storage wiring auto-lowers the storage version for a particular resource to whatever the compatibility version (n-1) server can handle.
> 

https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/server/storage/resource_encoding_config.go#L181-L196

I'm trying to understand why we can't directly put the most stable to least stable ordering in the install files. We check every resource with the introduced k8s version and the n-1 version and only allow it be eligible if it was introduced equal to or before the n-1 compat version. Then the first eligible version is selected based on the priority defined in `install.go`. Do we break something at the moment when we do that?

> Ideally, we could just put the most stable to least stable order in static install.go files, but that may introduce new instant-break situations when we create new release branches and master gets a next-minor development tag compared to the release branch, but must still pass tests for the previous minor

I think the scope of the etcd storage test is limited to integration testing? We still use the `DefaultKubeBinaryVersion` constant that is manually bumped, so there shouldn't be any test breakage there. If there are other tests that check the storage version, then yes those are likely to break, though I'm not aware of any other tests we need to manually bump/change on every release?

pohly replied:
https://github.com/kubernetes/kubernetes/issues/133131

TODO comment queued for next branch update.

liggitt commented on pkg/apis/resource/v1/conversion.go:
```diff
diff -u -r pkg/apis/resource/v1beta2/conversion.go pkg/apis/resource/v1/conversion.go
--- pkg/apis/resource/v1beta2/conversion.go	2025-07-02 12:33:32.498873356 -0400
+++ pkg/apis/resource/v1/conversion.go	2025-07-22 13:22:58.641742252 -0400
@@ -14,12 +14,12 @@
 limitations under the License.
 */
 
-package v1beta2
+package v1
 
 import (
 	"fmt"
 
-	resourceapi "k8s.io/api/resource/v1beta2"
+	resourceapi "k8s.io/api/resource/v1"
 	"k8s.io/apimachinery/pkg/runtime"
 )
```

liggitt commented on pkg/apis/resource/v1/doc.go:
```diff
diff -u -r pkg/apis/resource/v1beta2/doc.go pkg/apis/resource/v1/doc.go
--- pkg/apis/resource/v1beta2/doc.go	2025-07-02 12:33:32.499616854 -0400
+++ pkg/apis/resource/v1/doc.go	2025-07-22 13:22:58.642257044 -0400
@@ -15,9 +15,9 @@
 */
 
 // +k8s:conversion-gen=k8s.io/kubernetes/pkg/apis/resource
-// +k8s:conversion-gen-external-types=k8s.io/api/resource/v1beta2
+// +k8s:conversion-gen-external-types=k8s.io/api/resource/v1
 // +k8s:defaulter-gen=TypeMeta
-// +k8s:defaulter-gen-input=k8s.io/api/resource/v1beta2
+// +k8s:defaulter-gen-input=k8s.io/api/resource/v1
 
-// Package v1beta2 is the v1beta2 version of the resource API.
-package v1beta2
+// Package v1 is the v1 version of the resource API.
+package v1
```

liggitt commented on pkg/apis/resource/v1/register.go:
```diff
diff -u -r pkg/apis/resource/v1beta2/register.go pkg/apis/resource/v1/register.go
--- pkg/apis/resource/v1beta2/register.go	2025-07-02 12:33:32.499804229 -0400
+++ pkg/apis/resource/v1/register.go	2025-07-22 13:22:58.642432127 -0400
@@ -14,15 +14,15 @@
 limitations under the License.
 */
 
-package v1beta2
+package v1
 
 import (
-	"k8s.io/api/resource/v1beta2"
+	"k8s.io/api/resource/v1"
 	"k8s.io/apimachinery/pkg/runtime/schema"
 )
 
 var (
-	localSchemeBuilder = &v1beta2.SchemeBuilder
+	localSchemeBuilder = &v1.SchemeBuilder
 	AddToScheme        = localSchemeBuilder.AddToScheme
 )
 
@@ -38,7 +38,7 @@
 const GroupName = "resource.k8s.io"
 
 // SchemeGroupVersion is group version used to register these objects
-var SchemeGroupVersion = schema.GroupVersion{Group: GroupName, Version: "v1beta2"}
+var SchemeGroupVersion = schema.GroupVersion{Group: GroupName, Version: "v1"}
 
 // Resource takes an unqualified resource and returns a Group qualified GroupResource
 func Resource(resource string) schema.GroupResource {
```

liggitt commented on staging/src/k8s.io/api/resource/v1/devicetaint.go:
```diff
diff -u -r staging/src/k8s.io/api/resource/{v1beta2,v1}/devicetaint.go
--- staging/src/k8s.io/api/resource/v1beta2/devicetaint.go	2025-07-02 12:33:32.708155787 -0400
+++ staging/src/k8s.io/api/resource/v1/devicetaint.go	2025-07-22 13:22:58.744536652 -0400
@@ -14,7 +14,7 @@
 limitations under the License.
 */
 
-package v1beta2
+package v1
 
 import "fmt"
```

liggitt commented on staging/src/k8s.io/api/resource/v1/doc.go:
```diff
diff -u -r staging/src/k8s.io/api/resource/{v1beta2,v1}/doc.go
--- staging/src/k8s.io/api/resource/v1beta2/doc.go	2025-07-02 12:33:32.709076202 -0400
+++ staging/src/k8s.io/api/resource/v1/doc.go	2025-07-22 13:22:58.744643528 -0400
@@ -20,5 +20,5 @@
 // +k8s:prerelease-lifecycle-gen=true
 // +groupName=resource.k8s.io
 
-// Package v1beta2 is the v1beta2 version of the resource API.
-package v1beta2
+// Package v1 is the v1 version of the resource API.
+package v1
```

liggitt commented on staging/src/k8s.io/api/resource/v1/register.go:
```diff
diff -u -r staging/src/k8s.io/api/resource/{v1beta2,v1}/register.go
--- staging/src/k8s.io/api/resource/v1beta2/register.go	2025-07-02 12:33:32.710400491 -0400
+++ staging/src/k8s.io/api/resource/v1/register.go	2025-07-22 13:22:58.747091613 -0400
@@ -14,7 +14,7 @@
 limitations under the License.
 */
 
-package v1beta2
+package v1
 
 import (
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
@@ -26,7 +26,7 @@
 const GroupName = "resource.k8s.io"
 
 // SchemeGroupVersion is group version used to register these objects
-var SchemeGroupVersion = schema.GroupVersion{Group: GroupName, Version: "v1beta2"}
+var SchemeGroupVersion = schema.GroupVersion{Group: GroupName, Version: "v1"}
 
 // Resource takes an unqualified resource and returns a Group qualified GroupResource
 func Resource(resource string) schema.GroupResource {
```

liggitt commented on staging/src/k8s.io/api/resource/v1/types.go:
```diff
--- staging/src/k8s.io/api/resource/v1beta2/types.go	2025-07-22 13:22:58.754939205 -0400
+++ staging/src/k8s.io/api/resource/v1/types.go	2025-07-22 13:22:58.747328447 -0400
@@ -14,7 +14,7 @@
 limitations under the License.
 */
 
-package v1beta2
+package v1
 
 import (
 	v1 "k8s.io/api/core/v1"
@@ -36,7 +36,7 @@
 // +genclient
 // +genclient:nonNamespaced
 // +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
-// +k8s:prerelease-lifecycle-gen:introduced=1.33
+// +k8s:prerelease-lifecycle-gen:introduced=1.34
 
 // ResourceSlice represents one or more resources in a pool of similar resources,
 // managed by a common driver. A pool may span more than one ResourceSlice, and exactly how many
@@ -500,7 +500,7 @@
 )
 
 // +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
-// +k8s:prerelease-lifecycle-gen:introduced=1.33
+// +k8s:prerelease-lifecycle-gen:introduced=1.34
 
 // ResourceSliceList is a collection of ResourceSlices.
 type ResourceSliceList struct {
@@ -515,7 +515,7 @@
 
 // +genclient
 // +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
-// +k8s:prerelease-lifecycle-gen:introduced=1.33
+// +k8s:prerelease-lifecycle-gen:introduced=1.34
 
 // ResourceClaim describes a request for access to resources in the cluster,
 // for use by workloads. For example, if a workload needs an accelerator device
@@ -1314,7 +1314,7 @@
 )
 
 // +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
-// +k8s:prerelease-lifecycle-gen:introduced=1.33
+// +k8s:prerelease-lifecycle-gen:introduced=1.34
 
 // ResourceClaimList is a collection of claims.
 type ResourceClaimList struct {
@@ -1330,7 +1330,7 @@
 // +genclient
 // +genclient:nonNamespaced
 // +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
-// +k8s:prerelease-lifecycle-gen:introduced=1.33
+// +k8s:prerelease-lifecycle-gen:introduced=1.34
 
 // DeviceClass is a vendor- or admin-provided resource that contains
 // device configuration and selectors. It can be referenced in
@@ -1387,7 +1387,7 @@
 }
 
 // +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
-// +k8s:prerelease-lifecycle-gen:introduced=1.33
+// +k8s:prerelease-lifecycle-gen:introduced=1.34
 
 // DeviceClassList is a collection of classes.
 type DeviceClassList struct {
@@ -1402,7 +1402,7 @@
 
 // +genclient
 // +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
-// +k8s:prerelease-lifecycle-gen:introduced=1.33
+// +k8s:prerelease-lifecycle-gen:introduced=1.34
 
 // ResourceClaimTemplate is used to produce ResourceClaim objects.
 //
@@ -1437,7 +1437,7 @@
 }
 
 // +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
-// +k8s:prerelease-lifecycle-gen:introduced=1.33
+// +k8s:prerelease-lifecycle-gen:introduced=1.34
 
 // ResourceClaimTemplateList is a collection of claim templates.
 type ResourceClaimTemplateList struct {
```

liggitt commented on plugin/pkg/auth/authorizer/rbac/bootstrappolicy/testdata/cluster-roles.yaml:
```diff
@@ -863,6 +884,60 @@ items:
+    resources:
+    - pods/finalizers
+    verbs:
+    - get
+    - list
+    - patch
+    - update
+    - watch
```
pre-existing, but only the "update" verb is meaningful for this, might as well prune the verb set to just "update" when building the kubeSchedulerRules for this

pohly replied:
Fixed with a cleanup commit "DRA RBAC: fix kube-scheduler bootstrap policy", then updating the following "DRA: graduate DynamicResourceAllocation feature to GA" commit accordingly.

liggitt commented on plugin/pkg/auth/authorizer/rbac/bootstrappolicy/testdata/cluster-roles.yaml:
```diff
@@ -863,6 +884,60 @@ items:
+    - get
+    - list
+    - watch
+  - apiGroups:
+    - resource.k8s.io
+    resources:
+    - devicetaintrules
+    verbs:
+    - get
+    - list
+    - watch
```
pre-existing, but this bit in building kubeSchedulerRules should be guarded on the DRADeviceTaints feature, this shouldn't be showing up in default RBAC policy yet

liggitt reviewed (commented):
API promotion from v1beta2 to v1 lgtm

I noticed a couple existing things to tweak with the default RBAC policy around DRA stuff, authorization bits lgtm otherwise

I didn't review the kubelet / scheduler implementation changes, that needs lgtm / review from approvers for those packages

pohly committed:
DRA RBAC: fix kube-scheduler bootstrap policy

The pods/finalizer permission can be restricted to just updates because that is
all that matters.

The DeviceTaints rules were under the wrong feature gate check (copy-and-paste)
and must remain disabled when DRA itself becomes enabled.

nojnhuh reviewed (approved):
/lgtm

klueska commented:
/approve
For the kubelet side of things

nojnhuh reviewed (approved):
/lgtm

liggitt commented:
looks like the `pkg/scheduler/framework/autoscaler_contract/OWNERS` package needs an approval for the import change from v1beta2 → v1

that package looks like it is carved out from top-level approvals

/cc @kubernetes/sig-autoscaling-api-reviews

liggitt commented:
/approve
for API and authorization bits

nojnhuh reviewed (commented):
/lgtm

klueska commented:
I'll do the honors

/lgtm

pohly commented:
/hold cancel

We are working on stabilizing the windows-master and new DRAAdminAccess E2E metric test. Let's move forward here.

pacoxu commented:
/test pull-kubernetes-conformance-kind-ga-only-parallel
/test pull-kubernetes-node-e2e-containerd
/test pull-kubernetes-e2e-gce
/test pull-kubernetes-e2e-kind
/test pull-kubernetes-cmd
/test pull-kubernetes-integration
/test pull-kubernetes-typecheck
/test pull-kubernetes-unit
/test pull-kubernetes-verify

pacoxu commented:
/hold
to see if the CI failure needs to be fixed.
```
=== Failed
=== FAIL: k8s.io/dynamic-resource-allocation/structured  (0.00s)
FAIL	k8s.io/dynamic-resource-allocation/structured [build failed]

=== FAIL: k8s.io/dynamic-resource-allocation/structured/internal/experimental  (0.00s)
FAIL	k8s.io/dynamic-resource-allocation/structured/internal/experimental [build failed]

```

nojnhuh commented on test/e2e_dra/upgradedowngrade_test.go:
```diff
@@ -287,7 +287,7 @@ var _ = ginkgo.Describe("DRA upgrade/downgrade", func() {
 		failClaimTemplate.Name = "fail-template"
 		failClaimTemplate.Namespace = namespace.Name
-		_, err = f.ClientSet.ResourceV1beta2().ResourceClaimTemplates(namespace.Name).Create(ctx, failClaimTemplate, metav1.CreateOptions{})
+		_, err = f.ClientSet.ResourceV1().ResourceClaimTemplates(namespace.Name).Create(ctx, failClaimTemplate, metav1.CreateOptions{})
 		if err == nil {
 			tCtx.Fatalf("admin access template should fail in regular namespace but succeeded")
 		}
```
Since this part of the test is before an upgrade, the error that occurs here is [`the server could not find the requested resource`](https://prow.k8s.io/view/gs/kubernetes-ci-logs/pr-logs/pull/132706/pull-kubernetes-dra-integration/1948030673033891840#1:build-log.txt%3A1510) where it should probably stay something like it was before, [`Forbidden: admin access to devices requires the `resource.k8s.io/admin-access: true` label on the containing namespace`](https://prow.k8s.io/view/gs/kubernetes-ci-logs/pr-logs/pull/133110/pull-kubernetes-dra-integration/1947732259461140480#1:build-log.txt%3A1510).

Since that difference isn't causing the test to fail, I could address that in #132997 if that makes things easier.

nojnhuh commented:
CI is red since #133147 merged. Copying `{allocator,pools}_incubating.go` to experimental should get things in better shape.

pohly commented:
> CI is red since https://github.com/kubernetes/kubernetes/pull/133147 merged.

The additional file on master now also needs to use the V1 API. Fixed and pushed.

pohly commented:
/retest

klueska commented:
/lgtm

pacoxu commented:
/test pull-kubernetes-e2e-gce
